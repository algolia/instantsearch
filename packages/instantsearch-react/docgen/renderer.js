import marked from 'marked';
import {runMode} from 'codemirror/addon/runmode/runmode.node';
import 'codemirror/mode/jsx/jsx';
import escape from 'escape-html';

const highlight = source => {
  let output = '';
  // Even though highlight accepts a lang parameter, force to JSX for now.
  runMode(source, 'jsx', (text, style) => {
    text = escape(text);
    if (!style) {
      output += text;
      return;
    }
    const className = style.split(' ').map(s => `cm-${s}`).join(' ');
    output += `<span class="${className}">${text}</span>`;
  });
  return output;
};

const blockTheme = 'cm-s-default';
const spanTheme = 'cm-s-default';
const renderer = new marked.Renderer();
renderer.code = (code, lang) => {
  return `<pre class="CodeMirror ${blockTheme}"><code>${highlight(code, lang)}</code></pre>`;
};
renderer.codespan = code => {
  return `<code class="CodeMirror ${spanTheme}">${code}</code>`;
};

export default renderer;
