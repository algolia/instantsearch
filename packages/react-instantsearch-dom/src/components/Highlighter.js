import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

export const Highlight = ({
  cx,
  value,
  highlightedTagName,
  isHighlighted,
  nonHighlightedTagName,
}) => {
  const TagName = isHighlighted ? highlightedTagName : nonHighlightedTagName;
  const className = isHighlighted ? 'highlighted' : 'nonHighlighted';
  return <TagName className={cx(className)}>{value}</TagName>;
};

Highlight.propTypes = {
  cx: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  isHighlighted: PropTypes.bool.isRequired,
  highlightedTagName: PropTypes.string.isRequired,
  nonHighlightedTagName: PropTypes.string.isRequired,
};

const Highlighter = ({
  cx,
  hit,
  attribute,
  highlight,
  highlightProperty,
  tagName,
  nonHighlightedTagName,
  separator,
  className,
}) => {
  const parsedHighlightedValue = highlight({
    hit,
    attribute,
    highlightProperty,
  });

  return (
    <span className={classNames(cx(''), className)}>
      {parsedHighlightedValue.map((item, i) => {
        if (Array.isArray(item)) {
          const isLast = i === parsedHighlightedValue.length - 1;
          return (
            <span key={i}>
              {item.map((element, index) => (
                <Highlight
                  cx={cx}
                  key={index}
                  value={element.value}
                  highlightedTagName={tagName}
                  nonHighlightedTagName={nonHighlightedTagName}
                  isHighlighted={element.isHighlighted}
                />
              ))}
              {!isLast && <span className={cx('separator')}>{separator}</span>}
            </span>
          );
        }

        return (
          <Highlight
            cx={cx}
            key={i}
            value={item.value}
            highlightedTagName={tagName}
            nonHighlightedTagName={nonHighlightedTagName}
            isHighlighted={item.isHighlighted}
          />
        );
      })}
    </span>
  );
};

Highlighter.propTypes = {
  cx: PropTypes.func.isRequired,
  hit: PropTypes.object.isRequired,
  attribute: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.string,
  ]).isRequired,
  highlight: PropTypes.func.isRequired,
  highlightProperty: PropTypes.string.isRequired,
  tagName: PropTypes.string,
  nonHighlightedTagName: PropTypes.string,
  className: PropTypes.string,
  separator: PropTypes.node,
};

Highlighter.defaultProps = {
  tagName: 'em',
  nonHighlightedTagName: 'span',
  className: '',
  separator: ', ',
};

export default Highlighter;
