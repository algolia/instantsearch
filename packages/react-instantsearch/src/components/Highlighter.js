import React from 'react';

export default function Highlighter({hit, attributeName, highlight, highlightProperty}) {
  const parsedHighlightedValue = highlight({hit, attributeName, highlightProperty});
  const reactHighlighted = parsedHighlightedValue.map((v, i) => {
    const key = `split-${i}-${v.value}`;
    if (!v.isHighlighted) {
      return <span key={key} className="ais-Highlight__nonHighlighted">{v.value}</span>;
    }
    return <em key={key} className="ais-Highlight__highlighted">{v.value}</em>;
  });
  return <span className="ais-Highlight">{reactHighlighted}</span>;
}

Highlighter.propTypes = {
  hit: React.PropTypes.object.isRequired,
  attributeName: React.PropTypes.string.isRequired,
  highlight: React.PropTypes.func.isRequired,
  highlightProperty: React.PropTypes.string.isRequired,
};
