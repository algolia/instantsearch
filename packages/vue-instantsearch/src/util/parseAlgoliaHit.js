// copied from React InstantSearch
import { getPropertyByPath } from 'instantsearch.js/es/lib/utils';

import { unescape } from '../util/unescape';

const TAG_PLACEHOLDER = {
  highlightPreTag: '__ais-highlight__',
  highlightPostTag: '__/ais-highlight__',
};

/**
 * Parses an highlighted attribute into an array of objects with the string value, and
 * a boolean that indicated if this part is highlighted.
 *
 * @param {string} preTag - string used to identify the start of an highlighted value
 * @param {string} postTag - string used to identify the end of an highlighted value
 * @param {string} highlightedValue - highlighted attribute as returned by Algolia highlight feature
 * @return {object[]} - An array of {value: string, isHighlighted: boolean}.
 */
function parseHighlightedAttribute({ preTag, postTag, highlightedValue = '' }) {
  const splitByPreTag = highlightedValue.split(preTag);
  const firstValue = splitByPreTag.shift();
  const elements =
    firstValue === '' ? [] : [{ value: firstValue, isHighlighted: false }];

  if (postTag === preTag) {
    let isHighlighted = true;
    splitByPreTag.forEach((split) => {
      elements.push({ value: split, isHighlighted });
      isHighlighted = !isHighlighted;
    });
  } else {
    splitByPreTag.forEach((split) => {
      const splitByPostTag = split.split(postTag);

      elements.push({
        value: splitByPostTag[0],
        isHighlighted: true,
      });

      if (splitByPostTag[1] !== '') {
        elements.push({
          // Vue removes nodes which are just a single space (vuejs/vue#9208),
          // we replace this by two spaces, which does not have an impact,
          // unless someone would have `white-space: pre` on the highlights
          value: splitByPostTag[1] === ' ' ? '  ' : splitByPostTag[1],
          isHighlighted: false,
        });
      }
    });
  }

  return elements;
}

/**
 * Find an highlighted attribute given an `attribute` and an `highlightProperty`, parses it,
 * and provided an array of objects with the string value and a boolean if this
 * value is highlighted.
 *
 * In order to use this feature, highlight must be activated in the configuration of
 * the index. The `preTag` and `postTag` attributes are respectively highlightPreTag and
 * highlightPostTag in Algolia configuration.
 *
 * @param {string} preTag - string used to identify the start of an highlighted value
 * @param {string} postTag - string used to identify the end of an highlighted value
 * @param {string} highlightProperty - the property that contains the highlight structure in the results
 * @param {string} attribute - the highlighted attribute to look for
 * @param {object} hit - the actual hit returned by Algolia.
 * @return {object[]} - An array of {value: string, isHighlighted: boolean}.
 */
export function parseAlgoliaHit({
  preTag = TAG_PLACEHOLDER.highlightPreTag,
  postTag = TAG_PLACEHOLDER.highlightPostTag,
  highlightProperty,
  attribute,
  hit,
}) {
  if (!hit) throw new Error('`hit`, the matching record, must be provided');

  const highlightObject =
    getPropertyByPath(hit[highlightProperty], attribute) || {};

  if (Array.isArray(highlightObject)) {
    return highlightObject.map((item) =>
      parseHighlightedAttribute({
        preTag,
        postTag,
        highlightedValue: unescape(item.value),
      })
    );
  }

  return parseHighlightedAttribute({
    preTag,
    postTag,
    highlightedValue: unescape(highlightObject.value),
  });
}
