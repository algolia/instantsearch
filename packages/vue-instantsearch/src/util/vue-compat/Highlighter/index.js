export { default } from './index-vue2';
