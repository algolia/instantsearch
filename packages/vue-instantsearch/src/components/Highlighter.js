import { createHighlightComponent } from 'instantsearch-ui-components';
import {
  getHighlightedParts,
  getPropertyByPath,
  unescape,
} from 'instantsearch.js/es/lib/utils';

import { createElement, Fragment } from '../util/pragma';

const Highlight = createHighlightComponent({ createElement, Fragment });

export default {
  name: 'AisHighlighter',
  props: {
    hit: {
      type: Object,
      required: true,
    },
    attribute: {
      type: String,
      required: true,
    },
    highlightedTagName: {
      type: String,
      default: 'mark',
    },
    suit: {
      type: Function,
      required: true,
    },
    highlightProperty: {
      type: String,
      required: true,
    },
    preTag: {
      type: String,
      required: true,
    },
    postTag: {
      type: String,
      required: true,
    },
  },
  render() {
    const property =
      getPropertyByPath(this.hit[this.highlightProperty], this.attribute) || [];
    const properties = Array.isArray(property) ? property : [property];

    const parts = properties.map((singleValue) =>
      getHighlightedParts(unescape(singleValue.value || '')).map(
        ({ value, isHighlighted }) => ({
          // We have to do this because Vue gets rid of TextNodes with a single white space
          value: value === ' ' ? '  ' : value,
          isHighlighted,
        })
      )
    );

    return createElement(Highlight, {
      classNames: {
        root: this.suit(),
        highlighted: this.suit('highlighted'),
      },
      highlightedTagName: this.highlightedTagName,
      nonHighlightedTagName: Fragment,
      parts,
    });
  },
};
