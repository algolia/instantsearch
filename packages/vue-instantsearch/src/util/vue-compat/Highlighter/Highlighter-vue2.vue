<template>
  <span :class="suit()">
    <component
      v-for="({ value, isHighlighted }, index) in parsedHighlights"
      :class="[isHighlighted && suit('highlighted')]"
      :key="index"
      :is="isHighlighted ? highlightedTagName : TextNode"
      >{{ value }}</component
    >
  </span>
</template>

<script>
import { parseAlgoliaHit } from '../../parseAlgoliaHit';

const TextNode = {
  render() {
    return this.$slots.default;
  },
};

export default {
  name: 'AisHighlighter',
  props: {
    hit: {
      type: Object,
      required: true,
    },
    attribute: {
      type: String,
      required: true,
    },
    highlightedTagName: {
      type: String,
      default: 'mark',
    },
    suit: {
      type: Function,
      required: true,
    },
    highlightProperty: {
      type: String,
      required: true,
    },
    preTag: {
      type: String,
      required: true,
    },
    postTag: {
      type: String,
      required: true,
    },
  },
  data() {
    return { TextNode };
  },
  computed: {
    parsedHighlights() {
      return parseAlgoliaHit({
        attribute: this.attribute,
        hit: this.hit,
        highlightProperty: this.highlightProperty,
        preTag: this.preTag,
        postTag: this.postTag,
      });
    },
  },
};
</script>
