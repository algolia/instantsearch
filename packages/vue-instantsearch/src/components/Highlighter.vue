<template>
  <span :class="suit()">
    <component
      v-for="({ value, isHighlighted }, index) in parsedHighlights"
      :class="[isHighlighted && suit('highlighted')]"
      :key="index"
      :is="isHighlighted ? highlightedTagName : TextNode"
      >{{ value }}</component
    >
  </span>
</template>

<script>
import { parseAlgoliaHit } from '../util/parseAlgoliaHit';
import { getDefaultSlot, renderCompat, isVue3 } from '../util/vue-compat';

const TextNode = isVue3
  ? (props, context) => context.slots.default()
  : {
      render: renderCompat(function () {
        return getDefaultSlot(this);
      }),
    };

export default {
  name: 'AisHighlighter',
  props: {
    hit: {
      type: Object,
      required: true,
    },
    attribute: {
      type: String,
      required: true,
    },
    highlightedTagName: {
      type: String,
      default: 'mark',
    },
    suit: {
      type: Function,
      required: true,
    },
    highlightProperty: {
      type: String,
      required: true,
    },
    preTag: {
      type: String,
      required: true,
    },
    postTag: {
      type: String,
      required: true,
    },
  },
  data() {
    return { TextNode };
  },
  computed: {
    parsedHighlights() {
      return parseAlgoliaHit({
        attribute: this.attribute,
        hit: this.hit,
        highlightProperty: this.highlightProperty,
        preTag: this.preTag,
        postTag: this.postTag,
      });
    },
  },
};
</script>
