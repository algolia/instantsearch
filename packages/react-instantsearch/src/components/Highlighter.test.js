/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';

import Highlighter from './Highlighter';
import parseAlgoliaHit from '../core/highlight';

describe('Highlighter', () => {
  it('parses an highlighted attribute of hit object', () => {
    const hitFromAPI = {
      objectID: 0,
      deep: {attribute: {value: 'awesome highlighted hit!'}},
      _highlightProperty: {
        deep: {
          attribute: {
            value: {
              value: 'awesome <ais-highlight>hi</ais-highlight>ghlighted <ais-highlight>hi</ais-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = ({hit, attributeName, highlightProperty}) => parseAlgoliaHit({
      preTag: '<ais-highlight>',
      postTag: '</ais-highlight>',
      attributeName,
      hit,
      highlightProperty,
    });

    const tree = renderer.create(
      <Highlighter attributeName="deep.attribute.value"
                   hit={hitFromAPI}
                   highlight={highlight}
                   highlightProperty="_highlightProperty"/>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
