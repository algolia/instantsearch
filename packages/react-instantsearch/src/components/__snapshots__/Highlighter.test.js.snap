// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlighter parses an highlighted attribute of hit object 1`] = `
<span
  className="ais-Highlight"
>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    awesome 
  </span>
  <em
    className="ais-Highlight__highlighted"
  >
    hi
  </em>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    ghlighted 
  </span>
  <em
    className="ais-Highlight__highlighted"
  >
    hi
  </em>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    t!
  </span>
</span>
`;
